### PR TITLE
Support alternative normalization methods

### PR DESCRIPTION
**Pull request by MuhammedHasan**

### CODE REVIEW

File: src/grelu/model/blocks.py [hunk 1]
```diff
@@ -111,6 +111,8 @@ def __init__(
         pool_size: Optional[str] = None,
         dropout: float = 0.0,
         norm: bool = True,
+        norm_type='batch',
```
Comment (avantikalal) [inline src/grelu/model/blocks.py:114]: Please update the docstring to mention these new variables.

Comment (avantikalal) [reply src/grelu/model/blocks.py:114]: Repeat this comment for all other affected functions.

**Review by avantikalal — CHANGES_REQUESTED** [review]

**Comment by avantikalal** [comment]
@MuhammedHasan  Please fix the conflicts and docstrings, looks good otherwise.

**Comment by avantikalal** [comment]
@MuhammedHasan please update the docstrings for affected functions; then, we can merge this :)

File: src/grelu/model/blocks.py [hunk 2]
```diff
@@ -41,11 +41,15 @@ def __init__(
         dropout: float = 0.0,
         norm: bool = False,
         bias: bool = True,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/blocks.py:44]: add docstrings for dtype and device in this function

File: src/grelu/model/blocks.py [hunk 3]
```diff
@@ -107,10 +113,14 @@ def __init__(
         pool_size: Optional[str] = None,
         dropout: float = 0.0,
         norm: bool = True,
+        norm_type="batch",
+        norm_kwargs=None,
         residual: bool = False,
         order: str = "CDNRA",
         bias: bool = True,
         return_pre_pool: bool = False,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/blocks.py:122]: add docstrings for dtype and device in this function

File: src/grelu/model/blocks.py [hunk 4]
```diff
@@ -195,7 +222,8 @@ class ChannelTransformBlock(nn.Module):
         out_channels: Number of channels in the output
         act_func: Name of the activation function
         dropout: Dropout probability
-        norm: If True, apply batch norm
+        norm_type: Type of normalization to apply: 'batch', 'syncbatch', 'layer', 'instance' or None
+        norn_kwargs: Additional arguments to be passed to the normalization layer
```
Comment (avantikalal) [inline src/grelu/model/blocks.py:226]: norn -> norm

File: src/grelu/model/blocks.py [hunk 5]
```diff
@@ -211,7 +239,11 @@ def __init__(
         act_func: str = "relu",
         dropout: float = 0.0,
         order: str = "CDNA",
+        norm_type="batch",
+        norm_kwargs=None,
         if_equal: bool = False,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/blocks.py:245]: add docstrings for dtype and device in this function

File: src/grelu/model/blocks.py [hunk 6]
```diff
@@ -281,6 +328,8 @@ def __init__(
         act_func: str = "relu",
         pool_func: Optional[str] = None,
         pool_size: Optional[str] = None,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/blocks.py:331]: add docstrings for dtype and device in this function

File: src/grelu/model/blocks.py [hunk 7]
```diff
@@ -329,8 +382,17 @@ def __init__(self, in_channels: int, kernel_size: int) -> None:
             groups=in_channels,
             padding="same",
             bias=False,
+            dtype=dtype,
```
Comment (avantikalal) [inline src/grelu/model/blocks.py:385]: add docstrings for dtype and device in this function

File: src/grelu/model/blocks.py [hunk 8]
```diff
@@ -393,13 +455,23 @@ def __init__(
         dropout: float = 0.0,
         order: str = "CDNRA",
         crop_len: Union[int, str] = 0,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/blocks.py:458]: add docstrings for dtype and device in this function

File: src/grelu/model/blocks.py [hunk 9]
```diff
@@ -477,14 +551,30 @@ class FeedForwardBlock(nn.Module):
     """
 
     def __init__(
-        self, in_len: int, dropout: float = 0.0, act_func: str = "relu"
+        self,
+        in_len: int,
+        dropout: float = 0.0,
+        act_func: str = "relu",
+        **kwargs,
```
Comment (avantikalal) [inline src/grelu/model/blocks.py:558]: add docstring for **kwargs in this function

File: src/grelu/model/blocks.py [hunk 10]
```diff
@@ -523,6 +613,8 @@ def __init__(
         dropout: float = 0.0,
         act_func: str = "relu",
         norm: bool = False,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/blocks.py:616]: add docstrings for dtype and device in this function

File: src/grelu/model/blocks.py [hunk 11]
```diff
@@ -583,6 +679,8 @@ def __init__(
         pos_dropout: float,
         attn_dropout: float,
         ff_dropout: float,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/blocks.py:682]: add docstrings for dtype and device in this function

File: src/grelu/model/blocks.py [hunk 12]
```diff
@@ -650,6 +752,8 @@ def __init__(
         pos_dropout: float = 0.0,
         attn_dropout: float = 0.0,
         ff_dropout: float = 0.0,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/blocks.py:755]: add docstrings for dtype and device in this function

File: src/grelu/model/blocks.py [hunk 13]
```diff
@@ -692,23 +798,46 @@ class UnetBlock(nn.Module):
     Args:
         in_channels: Number of channels in the input
         y_in_channels: Number of channels in the higher-resolution representation.
+        norm_type: Type of normalization to apply: 'batch', 'syncbatch', 'layer', 'instance' or None
+        norn_kwargs: Additional arguments to be passed to the normalization layer
```
Comment (avantikalal) [inline src/grelu/model/blocks.py:802]: norn -> norm

File: src/grelu/model/blocks.py [hunk 14]
```diff
@@ -692,23 +798,46 @@ class UnetBlock(nn.Module):
...
     """
 
-    def __init__(self, in_channels: int, y_in_channels: int) -> None:
+    def __init__(
+        self,
+        in_channels: int,
+        y_in_channels: int,
+        norm_type="batch",
+        norm_kwargs=None,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/blocks.py:811]: add docstrings for dtype and device in this function

File: src/grelu/model/heads.py [hunk 15]
```diff
@@ -32,6 +32,8 @@ def __init__(
         act_func: Optional[str] = None,
         pool_func: Optional[str] = None,
         norm: bool = False,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/heads.py:35]: add docstrings for dtype and device in this function

File: src/grelu/model/heads.py [hunk 16]
```diff
@@ -80,6 +87,8 @@ def __init__(
         hidden_size: List[int] = [],
         norm: bool = False,
         dropout: float = 0.0,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/heads.py:90]: add docstrings for dtype and device in this function

File: src/grelu/model/layers.py [hunk 17]
```diff
@@ -297,6 +312,8 @@ def __init__(
         n_pos_features: int,
         pos_dropout: float = 0,
         attn_dropout: float = 0,
+        device=None,
```
Comment (avantikalal) [inline src/grelu/model/layers.py:315]: add docstrings for dtype and device in this function

File: src/grelu/model/models.py [hunk 18]
```diff
@@ -95,6 +95,8 @@ def __init__(
         crop_len: int = 0,
         # Final pool
         final_pool_func: str = "avg",
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/models.py:98]: add docstrings for dtype and device in this function

File: src/grelu/model/models.py [hunk 19]
```diff
@@ -154,6 +160,8 @@ def __init__(
         n_conv: int = 8,
         crop_len: Union[str, int] = "auto",
         final_pool_func: str = "avg",
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/models.py:163]: add docstrings for dtype and device in this function

File: src/grelu/model/models.py [hunk 20]
```diff
@@ -224,6 +236,8 @@ def __init__(
         gru_norm: bool = False,
         # Final pool
         final_pool_func: str = "avg",
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/models.py:239]: add docstrings for dtype and device in this function

File: src/grelu/model/models.py [hunk 21]
```diff
@@ -313,6 +331,8 @@ def __init__(
         ff_dropout: float = 0.0,
         # Final pool
         final_pool_func: str = "avg",
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/models.py:334]: add docstrings for dtype and device in this function

File: src/grelu/model/models.py [hunk 22]
```diff
@@ -393,6 +417,8 @@ def __init__(
         mlp_act_func: Optional[str] = "relu",
         mlp_hidden_size: List[int] = [8],
         dropout: float = 0.0,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/models.py:420]: add docstrings for dtype and device in this function

File: src/grelu/model/models.py [hunk 23]
```diff
@@ -471,6 +499,8 @@ def __init__(
         crop_len: int = 16,
         final_act_func: Optional[str] = None,
         final_pool_func: Optional[str] = "avg",
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/models.py:502]: add docstrings for dtype and device in this function

File: src/grelu/model/models.py [hunk 24]
```diff
@@ -573,12 +617,20 @@ def __init__(
         in_len: int,
         channels=300,
         kernel_size=19,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/models.py:620]: add docstrings for dtype and device in this function

File: src/grelu/model/models.py [hunk 25]
```diff
@@ -621,6 +673,8 @@ def __init__(
         # Head
         final_act_func: Optional[str] = None,
         final_pool_func: Optional[str] = "avg",
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/models.py:676]: add docstrings for dtype and device in this function

File: src/grelu/model/trunks/__init__.py [hunk 26]
```diff
@@ -57,6 +57,7 @@ def __init__(
         dropout: float = 0.0,
         # Crop
         crop_len: int = 0,
+        **kwargs,
```
Comment (avantikalal) [inline src/grelu/model/trunks/__init__.py:60]: add docstrings for **kwargs in this function

File: src/grelu/model/trunks/__init__.py [hunk 27]
```diff
@@ -117,6 +119,7 @@ def __init__(
         act_func: str = "relu",
         n_conv: int = 8,
         crop_len: Union[str, int] = "auto",
+        **kwargs,
```
Comment (avantikalal) [inline src/grelu/model/trunks/__init__.py:122]: add docstring for **kwargs in this function

File: src/grelu/model/trunks/__init__.py [hunk 28]
```diff
@@ -195,6 +199,8 @@ def __init__(
         n_gru: int = 1,
         dropout: float = 0.0,
         gru_norm: bool = False,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/trunks/__init__.py:202]: add docstrings for dtype and device in this function

File: src/grelu/model/trunks/borzoi.py [hunk 29]
```diff
@@ -19,6 +20,8 @@ class BorzoiConvTower(nn.Module):
         out_channels: Number of channels in the output
         kernel_size: Width of the convolutional kernel
         n_blocks: Number of convolutional/pooling blocks, including the stem
+        norm_type: Type of normalization to apply: 'batch', 'syncbatch', 'layer', 'instance' or None
+        norn_kwargs: Additional arguments to be passed to the normalization layer
```
Comment (avantikalal) [inline src/grelu/model/trunks/borzoi.py:24]: norn -> norm

File: src/grelu/model/trunks/borzoi.py [hunk 30]
```diff
@@ -29,6 +32,10 @@ def __init__(
         out_channels: int,
         kernel_size: int,
         n_blocks: int,
+        norm_type="batch",
+        norm_kwargs=None,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/trunks/borzoi.py:37]: add docstrings for dtype and device in this function

File: src/grelu/model/trunks/borzoi.py [hunk 31]
```diff
@@ -109,6 +140,10 @@ def __init__(
         n_pos_features: int,
         # Crop
         crop_len: int,
+        norm_type="batch",
+        norm_kwargs=None,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/trunks/borzoi.py:145]: add docstrings for dtype and device in this function

File: src/grelu/model/trunks/enformer.py [hunk 32]
```diff
@@ -21,6 +22,8 @@ def __init__(
         self,
         n_blocks: int,
         out_channels: int,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/trunks/enformer.py:25]: add docstrings for dtype and device in this function

File: src/grelu/model/trunks/enformer.py [hunk 33]
```diff
@@ -114,6 +123,8 @@ def __init__(
         attn_dropout: float,
         pos_dropout: float,
         ff_dropout: float,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/trunks/enformer.py:126]: add docstrings for dtype and device in this function

File: src/grelu/model/trunks/explainn.py [hunk 34]
```diff
@@ -69,6 +73,8 @@ def __init__(
         in_len: int,
         channels=300,
         kernel_size=19,
+        dtype=None,
```
Comment (avantikalal) [inline src/grelu/model/trunks/explainn.py:76]: add docstrings for dtype and device in this function

File: src/grelu/model/blocks.py [hunk 35]
```diff
@@ -82,7 +86,9 @@ class ConvBlock(nn.Module):
         pool_func: Name of the pooling function
         pool_size: Pooling width
         dropout: Dropout probability
-        norm: If True, apply batch norm
+        norm: If True, apply normalization layer
+        norm_type: Type of normalization to apply: 'batch', 'syncbatch', 'layer', 'instance' or None
+        norn_kwargs: Additional arguments to be passed to the normalization layer
```
Comment (avantikalal) [inline src/grelu/model/blocks.py:91]: norn -> norm

**Review by avantikalal — CHANGES_REQUESTED** [review]

**Comment by MuhammedHasan** [comment]
@avantikalal I think we can merge this branch now with docstrings.

**Review by avantikalal — APPROVED** [review]
